File: templates/matlab/bar_diverging.m
function fig = bar_diverging()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(2);
    labels = arrayfun(@(i)sprintf('factor %d',i),1:10,'UniformOutput',false);
    v = -5 + 10*rand(1, 10); [v, idx] = sort(v); labels = labels(idx);
    fig = figure; hold on;
    for i = 1:numel(v)
        if v(i) < 0, c = palette('cat',2); else, c = palette('cat',1); end
        barh(i, v(i), 'FaceColor', c, 'EdgeColor','none');
    end
    set(gca,'YTick',1:numel(v),'YTickLabel',labels);
    xline(0, 'k'); xlabel('effect'); title('Diverging bar'); grid on;
end

File: templates/python/bar_diverging.py
"""bar_diverging: 发散柱状（正负对比，tornado chart）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import cycle

def make_figure(labels=None, values=None, title='Diverging bar'):
    apply_theme()
    if labels is None:
        labels = [f'factor {i+1}' for i in range(10)]
        values = np.random.default_rng(2).uniform(-5, 5, 10)
    order = np.argsort(values)
    labels = [labels[i] for i in order]; values = values[order]
    colors = [cycle(1) if v < 0 else cycle(0) for v in values]
    fig, ax = plt.subplots()
    ax.barh(labels, values, color=colors)
    ax.axvline(0, color='k', linewidth=0.6)
    ax.set_xlabel('effect'); ax.set_title(title)
    ax.grid(True, axis='x', linestyle=':', alpha=0.5)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure()
    fig.savefig(__file__.replace('.py', '.png'), dpi=150)
